fix(VListItem): allow default color to change when inactive

Tofandel opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://vuetifyjs.com/getting-started/contributing

Testing and markup sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

<!-- We use conventional-changelog-angular for all commit structures -->
<!-- https://vuetifyjs.com/getting-started/contributing#commit-guidelines-w-commitizen -->


## Description
<!-- Describe your changes in detail -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #4213 or fixes #2312 -->
Fixes #9285 by removing the !important keyword in the vuetify style sheet ~~and adding a `base-color` prop~~

## Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
It is currently not possible to change the color of a v-list-item via an hex `color="#CCC"` prop, it is only possible to do it via a tricky css rule (needs to be very specific, and the selector with this rule is already too specific with an `!important` keyword, 

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | e2e | none -->
Visually

#### Before
![image](https://user-images.githubusercontent.com/6115458/120027427-a4eebe80-bff3-11eb-922c-22f4e6ae9178.png)

#### After
![image](https://user-images.githubusercontent.com/6115458/120026787-e3d04480-bff2-11eb-90f6-e7190659c474.png)


## Markup:
<!-- Information on how to setup your local development environment can be found here: -->
<!-- https://vuetifyjs.com/getting-started/contributing#setup-dev-environment -->

<!-- Paste markup for testing your change --->
<details>

```vue
<template>
      <v-list dense>
        <v-list-item-group v-model="item">
          <template v-for="(item, i) in items">
          <v-list-item
            :key="i+'0'"
            :value="0"
            v-bind="item"
          >
            <v-list-item-icon>
              <v-icon>settings</v-icon>
            </v-list-item-icon>
            <v-list-item-content>
              <v-list-item-title>normal {{ item }}</v-list-item-title>
            </v-list-item-content>
          </v-list-item>
          <v-list-item
            :key="i+'1'"
            :value="1"
            v-bind="item"
          >
            <v-list-item-icon>
              <v-icon>settings</v-icon>
            </v-list-item-icon>
            <v-list-item-content>
              <v-list-item-title>active {{ item }}</v-list-item-title>
            </v-list-item-content>
          </v-list-item>
          <v-list-item
            dark
            :key="i+'0d'"
            :value="0"
            v-bind="item"
          >
            <v-list-item-icon>
              <v-icon>settings</v-icon>
            </v-list-item-icon>
            <v-list-item-content>
              <v-list-item-title>normal dark {{ item }}</v-list-item-title>
            </v-list-item-content>
          </v-list-item>
          <v-list-item
            dark
            :key="i+'1d'"
            :value="1"
            v-bind="item"
          >
            <v-list-item-icon>
              <v-icon>settings</v-icon>
            </v-list-item-icon>
            <v-list-item-content>
              <v-list-item-title>active dark {{ item }}</v-list-item-title>
            </v-list-item-content>
          </v-list-item>
          </template>
        </v-list-item-group>
      </v-list>
</template>

<script>
export default {
data: () => ({
    item: 1,
    items: [
      { color: "#228822" },
      { color: 'amber' },
      { baseColor: "#228822" },
      { baseColor: 'amber' },
      {}
    ],
  }),
}
</script>
```
</details>

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [x] I added relevant changes to the documentation

Tofandel commented:
Should the normal icon's color change as well maybe?

KaelWD commented:
Yes, but this was intentional so inactive items aren't highlighted (see https://vuetify-lzjyu0p1p-vuetifyjs1.vercel.app/en/getting-started/installation/)
![image](https://user-images.githubusercontent.com/16421948/119959963-d3ff3300-bfe7-11eb-8493-1b1cea945eb5.png)
The text here is supposed to be black unless it's an active route.

Tofandel commented:
Okay let me see if I can fix that without an important keyword

Tofandel commented:
I think to solve this it would make sense to have a `color` prop and an `active-color` prop or to avoid a breaking change a `base-color` prop and `color` prop (which only applies when the list item is active)

Tofandel commented:
I removed the color class when the item is not active, which fixes the demonstrated issue

Also took the liberty of adding the base-color prop which is the prop that controls the color when it's not active (and the color prop defaults to it if its not set)

KaelWD commented:
This also changes icons to black when inactive:
![image](https://user-images.githubusercontent.com/16421948/120195821-85a99880-c262-11eb-9417-36cd975ce779.png)

![image](https://user-images.githubusercontent.com/16421948/120195835-87735c00-c262-11eb-8222-d88ad6fedd3c.png)

Tofandel commented:
I changed it to only inherit icon color if the base color is set

I also added a colorful example in the doc
![image](https://user-images.githubusercontent.com/6115458/120205961-e1f0c600-c22a-11eb-9384-3eefe96e2818.png)

johnleider reviewed (changes requested):
I think the naming is confusing. If we have a visual change based upon the state of an item, it should have a name that reflects that.

Personally, I'd rather have the parent dictate a **color** and **active-color** and leave `v-list-item` alone.

I'm actively working on v3 lists and I'll chime in more as I experiment more.

Tofandel commented:
I agree that adding a `color` and `active-color` prop to the v-list would be a good idea to apply it to all the v-list-items

But it's a must to also be able to change the color of some of the list item individually, where you can override this

And sadly on the v-list-item the naming can't be changed, it would be a big breaking change

KaelWD commented:
`active-color` can't be done in v2 without being a breaking change, I'm ok with this as a stopgap then change it again in v3. 

> I'd rather have the parent dictate a color and active-color and leave v-list-item alone.

There does need to be some way to change the colour of an individual item imo

KaelWD commented:
Hey @Tofandel, we discussed this in today's meeting and decided not to implement the new prop in v2 as it would be immediately deprecated. To fix #9285 in v2 we can accept removing the active class when not active though.

Tofandel commented:
So just the fix without the base-color prop?

KaelWD commented on packages/vuetify/src/components/VList/VListItem.ts:
```diff
@@ -35,6 +35,8 @@ interface options extends ExtractVue<typeof baseMixins> {
   isInMenu: boolean
   isInNav: boolean
+  genAttrs (): object
+  click (e: MouseEvent | KeyboardEvent): void
```
This shouldn't be necessary.

johnleider commented:
Can you take a look at `v-list` as of v3-alpha.5 and compare the styling. We went with the prop name **active-color**.

Tofandel commented:
@johnleider The doc site doesn't seem to include it there https://next.vuetifyjs.com/en/components/application/ any other way to check except building it locally? Did you also add a similar snippet in the doc like: https://github.com/vuetifyjs/vuetify/pull/13719/commits/54f8d4f26e92175d6f337d188c4fbf79edddcc78 ?

KaelWD commented:
@Tofandel I'm going to have to revert this, `color` no longer affects active route links: #14186

Feel free to open a new PR if you have a solution that works for this case as well.

Tofandel commented:
@KaelWD that seems weird, isn't `this.isActive` data true for router links? Because I swear this was working

Tofandel commented:
So yeah this.isActive is false even if the link is current is this normal? Isn't there something missing in https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/mixins/routable/index.ts

Tofandel commented:
One workaround is to check for the activeClass on the element instead of the isActive on every route change (or to set the isActive to true) because there doesn't seem to be any event or data available to save this externally in vue router..

Tofandel commented:
Well there you go.. It's already implemented https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/mixins/routable/index.ts#L153 The toggle function is empty and should be `this.isActive = true`

Tofandel commented:
@KaelWD So if you can do it quickly you can rerevert and change the line in the routable mixin, otherwise I'll make a new PR

Tofandel commented:
Indeed made a new PR, the onRouteChange should be called on mounted, as for the activeClass to set it back to false no need, you actually just need to call toggle, only if the class boolean is different than the isActive value, effectively trully reflecting isActive on every route change
The toggle also needs to be redefined in the VListItem because it's overriden by the groupable mixin
